wasm-test/src/lib.rs

```diff
@@ -0,0 +1,34 @@
+mod compile;
+mod environment;
+mod test_case;
+
+pub use compile::Compiler;
+pub use environment::TestEnvironment;
+pub use test_case::{Pass, Recipe, TestCase};
+
+use anyhow::{Context, Error};
+
+pub fn run_test_case(
+    compiler: &Compiler,
+    test_case: &TestCase,
+) -> Result<(), Error> {
+    let mut wasm = compiler
+        .instantiate(&test_case.name, &test_case.src)
+        .context("Unable to load the WASM module")?;
+    let mut env = TestEnvironment::default();
+
+    for pass in &test_case.recipe.passes {
+        env.setup(pass);
+
+        log::debug!("Polling \"{}\" at {:?}", test_case.name, env.elapsed);
+        log::trace!("Environment: {:?}", env);
```

Comment:
I know that trace is supposed to be finer grained than debug. What is your strategy for the different log levels?

Reply:
From my experience trying to debug complex interactions at work a strategy I came up with is to use `debug` for getting a general idea of what happens at a particular stage, but if you need to see *all* the internals (e.g. the exact state of every byte in the process image) then you'd switch the log level to `trace`.